### ISSUE TITLE
What should be done about tests that involve the PUF?

### ISSUE THREAD

**Issue by talumbau** [open-source-economics/Tax-Calculator#528]
A discussion was started in #527 about tests that would benefit from having the PUF. This issue is to discuss the matter more fully. I think there are at least two ways to handle the issue:

1. Create tests that require the PUF, then create a mechanism that would allow Travis CI to access the PUF and run the tests (e.g. have Travis CI download the PUF from S3 with every PR using credentials that we keep private). If we do this, any test may require the PUF.

2. Create tests that require the PUF, then scope the tests that run on Travis CI to exclude those tests. Developers posting a PR would be expected to run the tests that require the PUF, but Travis CI would only run the subset of tests that do not require the PUF.

If we do option 2, the scoping of the tests for Travis CI could be done via `pytest.mark`, or another mechanism like `pytest.skipif` as @martinholmer mentioned. `skipif` would skip the test even if the developer was "supposed" to run the test with the PUF, so perhaps `mark` would be better. With `pytest.mark` you just decorate the test with a marker that make sense to you:

```
@pytest.mark.requires_puf
def test_something_using_the_puf
    pass
```

Developers would continue to execute `py.test` as normal, but we would set up Travis CI to exclude this test, by changing the .travis.yml file to something like this:

```
py.test -m 'not requires_puf'
```
So, it's a more explicit statement of whether or not to run certain tests.

Thoughts?

cc @martinholmer @GoFroggyRun @MattHJensen @Amy-Xu

**Comment by martinholmer** [comment]
T.J. said in part:
> A discussion was started in #527 about tests that would benefit from having the PUF. This issue is to discuss the matter more fully. I think there are at least two ways to handle the issue: ...[snip]...

T.J. thanks for your useful discussion in issue #528, but maybe my suggestions were too vague to be clear.  Yes, it might be useful to have Travis-CI do puf.csv-input tests for those with the proper credentials, but my main interest is to have py.test run puf.csv-input tests on a credentialed developer's local computer **before** they submit a pull request for a branch that they have developed.  If only one capability was possible, it would seem to me that the ability for py.test to run puf.csv-input tests on a local computer (if the puf.csv file was available) would be the more valuable capability.  Of course, having both capabilities would be ideal.

Can either or both of your two ways to handle this issue be generalized to a local computer (that is, to have nothing to do with Travis-CI)?

cc @talumbau @MattHJensen @Amy-Xu @Amy-Xu

**Comment by talumbau** [comment]
I think there is something here that I don't understand, so I will just try to state what I understand as potential goals and how one could achieve those goals. I think we have established a culture well enough for people to have the habit of typing `py.test` and seeing the results prior to creating a PR. If we don't have an established culture of developers at least doing that much, then administrative controls of the type we are discussing in this issue are something that should be discussed only after creating a culture where testing prior to a push/opening a PR is the norm. As I said, I think we do have such a culture.

Potential Goal 1: force someone to run a test with the PUF before they submit a PR. 
If we can assume that someone would run `py.test` prior to submitting the PR, then we could just write tests that require the PUF. I would suggest marking them as such, using a decorator like `pytest.mark.requires_puf`. These tests can be written to assume the PUF has a particular filename and is in a particular location in the local development repo. We can disclose this info in the documentation. If a developer does not have the PUF, he/she will get a failed test. The advantage of this style is that the developer only has to keep doing the standard behavior (type `py.test` at the top of the repo prior to creating the PR). If someone wants to contribute and does not have access to the PUF, they must assert that they are not running the PUF-required tests, by typing `py.test -m 'not requires_puf'`, which we can also document.
 
Potential Goal 2: force the execution of a test with the PUF if the PUF is in a specific location in a local repository, otherwise skip the test.
I don't think this is a good goal because I think we should create testing structures that are explicit about what is being tested, rather than implicit. However, to accomplish this, one can use a `pytest.skip_if` decorator.

Achieving Goal 1 has some good up-sides: once a developer has the PUF, they can "set it and forget it". They test as normal and get PUF-enabled testing. If their development environment changes, they won't forget to test with the PUF, because they'll get test failures. A developer without the PUF can easily still contribute, but their behavior reflects that they are not running the full tests suite (i.e. they must type something like `-m 'not requires_puf'`.

Achieving Goal 2 means that developers with and w/o the PUF both only ever type the same thing (i.e. `py.test`), but it would require additional diligence on the part of the developer with the PUF to make sure that tests that require the PUF are actually running, since they could potentially get "skips" on the PUF tests if they have their environment set up incorrectly, even though they have the PUF on their machine.

Am I correctly characterizing the desired goals? Is there a further goal that I have missed?

**Comment by feenberg** [comment]
On Mon, 18 Jan 2016, T.J. Alumbaugh wrote:

> 
> I think there is something here that I don't understand, so I will just try to state
> what I understand as potential goals and how one could achieve those goals. I think we
> have established a culture well enough for people to have the habit of typing py.test
> and seeing the results prior to creating a PR. If we don't have an established culture
> of developers at least doing that much, then administrative controls of the type we
> are discussing in this issue are something that should be discussed only after
> creating a culture where testing prior to a push/opening a PR is the norm. As I said,
> I think we do have such a culture.

Is the file referred to above as the PUF the one from 2008, 2009 or 2010?

Not all potential contributors to TaxBrain will have that file, will they? 
Might they be depending on AEI to provide the data, and expect to use 
AEI's copy of the calculator after the PR is acted on? So shouldn't we 
offer something to use in place of the 2009 PUF?

dan

**Comment by martinholmer** [comment]
T.J. said on 18-Jan-2016:
> Am I correctly characterizing the desired goals? Is there a further goal that I have missed?

YES to your first question and NO to your second question.

I like very much the first method you suggest, which you describe as follows (with minor clarifying edits):
> If we can assume that someone would run ```py.test``` prior to submitting the PR, then we could just write tests that require the puf.csv file. I would suggest marking them as such, using a decorator like ```pytest.mark.requires_pufcsv.``` These tests can be written to assume the puf.csv file is in a particular location in the local development repo. We can disclose this info in the documentation. If a developer does not have the puf.csv file, he/she will get a failed test. The advantage of this style is that the developer only has to keep doing the standard behavior (type ```py.test``` at the top of the repo when developing a PR). If someone wants to contribute and does not have access to the puf.csv file, they must assert that they are not running the puf.csv-required tests, by typing ```py.test -m "not requires_pufcsv"```, which we can also document.

I have only one question.  Will Travis-CI alway execute the tests marked ```requires_pufcsv```?  I'm assuming the answer is YES, but just want to clarify.  In an earlier comment, you had described Travis-CI getting a copy of puf.csv from AWS S3 storage, if I remember correctly.

cc @talumbau @MattHJensen @feenberg @Amy-Xu @GoFroggyRun

**Comment by talumbau** [comment]
OK great. This seems like a reasonable path forward. In terms of Travis CI, we can have our choice: the easiest thing right now would be to make it so that Travis CI runs the tests in the same way a developer would run the test if he/she did not have the PUF.csv file. When we check in the first tests that require the puf.csv file, we would simply make a one-line change in the `.travis.yml` file so that Travis CI does not run that test. This seems like an OK first step because we would be no worse off than we are right now, since Travis CI does not run such tests at present. The idea of an automatic download using private credentials would be fairly easy to set up, but I'd want to check in with @MattHJensen first to see if the details are OK with how we have to control access to the PUF.

**Comment by talumbau** [comment]
> Is the file referred to above as the PUF the one from 2008, 2009 or 2010?
> 
> Not all potential contributors to TaxBrain will have that file, will they?
> Might they be depending on AEI to provide the data, and expect to use
> AEI's copy of the calculator after the PR is acted on? So shouldn't we
> offer something to use in place of the 2009 PUF?

@feenberg we do have a "simulated" PUF in the test suite right now, which is mainly used to make sure it's possible to call all of the functions and basically "smoke test" the calculator. The PUF file I'm referring to is the one we use in production, the latest version of which was created by @Amy-Xu (I believe it's from 2009 but Amy would know for sure). A truly useable "synthetic" PUF (that is statistically sound and representative of real tax data) seems like it would be a great resource for this community, but to my knowledge does not exist. So, I think we are stuck with the very incomplete (but useable) `tax_all1991_puf.gz` file at the top of the repo, and the proposals in this issue to work around the fact that some people will not have the actual PUF file.

**Comment by feenberg** [comment]
On Mon, 18 Jan 2016, T.J. Alumbaugh wrote:

>
>       Is the file referred to above as the PUF the one from 2008, 2009 or 2010?
>
>       Not all potential contributors to TaxBrain will have that file, will they?
>       Might they be depending on AEI to provide the data, and expect to use
>       AEI's copy of the calculator after the PR is acted on? So shouldn't we
>       offer something to use in place of the 2009 PUF?
> 
> @feenberg we do have a "simulated" PUF in the test suite right now, which is mainly
> used to make sure it's possible to call all of the functions and basically "smoke
> test" the calculator. The PUF file I'm referring to is the one we use in production,
> the latest version of which was created by @Amy-Xu (I believe it's from 2009 but Amy
> would know for sure). A truly useable "synthetic" PUF (that is statistically sound and
> representative of real tax data) seems like it would be a great resource for this

Eventually I will produce a file from the SCF that can work with the 
calculator. It will be representative of the population, but have only the 
22 variables used by online TAXSIM. So it won't be great for testing, but 
will be usable for simple scoring, such as changes in rates or exemptions.

dan

**Comment by Amy-Xu** [comment]
@feenberg We're using the CPS-matched 2009 PUF. 

The approach sounds very practical to me. At least for now, every developer has a copy of PUF.

**Comment by martinholmer** [comment]
T.J. said:
> OK great. This seems like a reasonable path forward. In terms of Travis CI, we can have our choice: the easiest thing right now would be to make it so that Travis CI runs the tests in the same way a developer would run the test if he/she did not have the PUF.csv file. When we check in the first test that requires the puf.csv file, we would simply make a one-line change in the .travis.yml file so that Travis CI does not run that test. This seems like an OK first step because we would be no worse off than we are right now, since Travis CI does not run such tests at present. The idea of an automatic download using private credentials would be fairly easy to set up, but I'd want to check in with @MattHJensen first to see if the details are OK with how we have to control access to the PUF.

Your Travis-CI strategy sounds quite sensible.  While you are checking with Matt about how to provide Travis-Ci with the puf.csv file in a way that meets IRS security requirements, I'll start by transforming the testagg.py and testmtr.py logic into pytests in the taxcalc/tests directory.

Do you think the best location for the optional puf.csv file is the top-level tax-calculator directory?  Matt told me that some developers keep it there even though it is not under Git version control. Would there be a better location?  I'm not familiar with the mechanics of Travis-CI and whether those mechanics would have any implications for where we tell the new pytests to look for the puf.csv file.

cc @talumbau @MattHJensen @feenberg @Amy-Xu @GoFroggyRun

**Comment by MattHJensen** [comment]
>  have Travis CI download the PUF from S3 with every PR using credentials that we keep private

The S3 PUF storage should be fine. I'm excited about the new testing strategy.

**Comment by martinholmer** [comment]
Matt said:
>> have Travis CI download the PUF from S3 with every PR using credentials that we keep private
>
> The S3 PUF storage should be fine. I'm excited about the new testing strategy.

Maybe I'm confused by the quote referenced by Matt, but don't we want Travis-CI to temporarily move the puf.csv file from AWS S3 storage for **every** pull request?  Why only for core developers who have local access to the puf.csv file?  If we don't have Travis-CI run the tests marked ```requires_pufcsv``` on every pull request, then how are we to know if a newcomer's pull request is sensible or not?

cc @MattHJensen @talumbau @feenberg @Amy-Xu @GoFroggyRun

**Comment by MattHJensen** [comment]
>but don't we want Travis-CI to temporarily move the puf.csv file from AWS S3 storage for every pull request?

Yes, sorry, that's what I was trying to convey. So Travis-CI would have the S3 credentials, but those credentials would not be shared with non-core developers.

**Comment by MattHJensen** [comment]
Is there a way to allow the core team to review a PR before triggering the PUF-based tests?

**Comment by martinholmer** [comment]
Matt said:
> Is there a way to allow the core team to review a PR before triggering the PUF-based tests?

Look at pull request #537 and give me your reactions.

**Comment by talumbau** [comment]
Matt said:

> Is there a way to allow the core team to review a PR before triggering the PUF-based tests?

No, not if the expectation is that Travis CI is running these tests. The "trigger" is the opening of a PR against the repo `github.com:open-source-economics/Tax-Calculator`

**Comment by MattHJensen** [comment]
Could the user add a requires_pufcsv test that discloses sensitive information?

Is there a way to block the contributor from adding requires_pufcsv tests?

**Comment by martinholmer** [comment]
Matt said:
> Could a user add a ``requires_pufcsv``` test that discloses sensitive information?

If a non-core developer proposed such a test, then I assume the core developers would reject the pull request.  If a developer with master merge privileges were to merge such a test, that merge could be reverted as you have described to me in the past.  I see what you are worrying about as less of a risk than giving the ```puf.csv``` file to the core development team, any one of whom, I suppose, could simply distribute the file across the Internet.  There would be no recovery from that event.

Matt also said:
> Is there a way to block the contributor from adding ```requires_pufcsv``` tests?

I think the answers to this question are the same as for your first question.

So, in summary, I don't see these ```requires_pufcsv``` tests as increasing the IRS PUF security risks above their current level.  Am I missing something?

cc @MattHJensen @talumbau @feenberg @Amy-Xu @GoFroggyRun

**Comment by MattHJensen** [comment]
@martinholmer, I could be confused, but I thought that if a contributor were to both add a test and tag that test with 'requires_pufcsv'  in tax-calculator/TESTING.md, then the test will be executed when the pull request is *opened* and the contributor will be able to see the result of that test in the Travis CI output. In other words, the contributor will see the test results before the core contributors can review what the new test is.

**Comment by martinholmer** [comment]
Matt said:
> @martinholmer, I could be confused, but I thought that if a contributor were to both add a test and tag that test with 'requires_pufcsv' [as described in] tax-calculator/TESTING.md, then the test will be executed when the pull request is opened and the contributor will be able to see the result of that test in the Travis CI output. In other words, the contributor will see the test results before the core contributors can review what the new test is.

Yes, I think you are correct about the mechanics: if the test wrote IRS/SOI PUF individual data to stdout and this user made it so the test failed, then the Travis-CI report would include all the information written to stdout.  This would be a particularly malevolent user, but I suppose we need to think about all eventualities no matter how unlikely.

A couple of approaches for dealing with this situation come to mind.  I don't know whether or not GitHub and Travis-CI have the capability to implement these approaches.  I guess T.J. could tell us whether or not these approaches are technically feasible.

Both approaches require having a list of GitHub trusted users for the Tax-Calculator repository.

First approach: when a pull request is submitted to the Tax-Calculator GitHub repo, Travis-CI would get the puf.csv file and execute the requires_pufcsv tests only if the person submitting the pull request is on that trusted list.  If not on that trusted list, Travis-CI would skip the requires_pufcsv tests.

Second approach: when a pull request is submitted to the Tax-Calculator GitHub repo, it would be rejected if any changes were made to the test_pufcsv.py file and if the user submitting the pull request was not on that trusted list.

After thinking a bit about these two approaches, I think the second one is preferable because it allows non-trusted users to contribute enhancements or fixes that have been fully tested.  The first approach is inferior because the core development team will have to execute the requires_pufcsv tests on their local computer every time a non-trusted user submits a pull request, which is likely to be very time consuming.

cc @MattHJensen @talumbau @feenberg @Amy-Xu @GoFroggyRun

**Comment by MattHJensen** [comment]
> Second approach: when a pull request is submitted to the Tax-Calculator GitHub repo, it would be rejected if any changes were made to the test_pufcsv.py file and if the user submitting the pull request was not on that trusted list.


This seems preferred as well. 

If it is not possible to have a trusted list, though, what about just keeping test_pufcsv.py outside of the repo, perhaps in a separate repo? This separate repo would not even need to be private, because core contributors would be able to review a PR to tax-calculator before deciding whether to merge the PR to the separate repo, and the test wouldn't get triggered automatically.

**Comment by martinholmer** [comment]
Matt said:
>> Second approach: when a pull request is submitted to the Tax-Calculator GitHub repo, it would be rejected if any changes were made to the test_pufcsv.py file and if the user submitting the pull request was not on that trusted list.
>
> This seems preferred as well.
>
> If it is not possible to have a trusted list, though, what about just keeping test_pufcsv.py outside of the repo, perhaps in a separate repo? This separate repo would not even need to be private, because core contributors would be able to review a PR to tax-calculator before deciding whether to merge the PR to the separate repo, and the test wouldn't get triggered automatically.

I'm not sure how what you're suggesting would solve the problem if the separate repository was public.  But maybe I'm confused about what you're suggesting.  And, even if something like this works from a security perspective, it sounds like a lot of extra work for the core developers.

But, it does appear as if we are agreed that the second approach using a trusted list of GitHub users (as described above), is the preferred approach.  We just need to have T.J. tell us how to implement that second approach.

cc @MattHJensen @talumbau @feenberg @Amy-Xu @GoFroggyRun

**Comment by MattHJensen** [comment]
> I'm not sure how what you're suggesting would solve the problem if the separate repository was public. 

The idea is that anyone could open a PR to the test_pufcsv_repo that modifies `test_pufcsv.py`, but the Tax-Calculator Travis CI would only point to the checked in version of `test_pufcsv.py` in the test_pufcsv_repo. Therefore, no test in the Tax-Calculator would be triggered until *after* a core developer had both reviewed the Tax-Calculator and test_pufcsv_repo PRs and merged the PR to the test_pufcsv_repo. A core contributor may also need to retrigger the Travis CI test for the Tax-Calculator PR after merging the test_pufcsv_repo PR.

As long as there aren't a lot of outside contributors adding tests that require the puf, this wouldn't be much additional work for anyone.

**Comment by talumbau** [comment]
Ok, I'm all caught up on this issue now. Here's some relevant info:

Reading through the issue made me aware of the security problem. It is correct to say that if Travis CI is set up to download the PUF, then someone could write a test that prints out the output of the PUF and does any manner of things with the file. It doesn't matter if they are modifying a particular file (like `test_pufcsv.py`), they could just add such a test anywhere they wanted. Basically, Travis CI allows for the execution of arbitrary code when a PR is posted, so if we are downloading the PUF with Travis CI, the user's arbitrary code can read it.

I like the idea of a `test_puf_repo`. This could be a private repo that actually downloads the PUF and runs PUF related tests. What we want is some kind of mechanism such that, when a PR is posted to this repo, a core developer could easily trigger some PUF related tests against that PR, after having seen that the offered code is genuine code and not something malicious.

We can still merge https://github.com/open-source-economics/Tax-Calculator/pull/537 while discussing this issue, since by merging Martin's PR we are asserting that developers by default will run PUF related tests, but Travis CI will not. This seems to me to be an OK state.

**Comment by MattHJensen** [comment]
> This is not the typical Travis CI use case, so I'll have to look into what can be done.

@talumbau, thanks for the comments and offer to look into this. We'll wait to see what you come up with.

**Comment by martinholmer** [comment]
@talumbau,  Should you close Tax-Calculator issue #528, now that we have --- at your suggestion --- incorporated two tests that are marked ```requires_puf_csv``` into the py.test suite?

**Comment by talumbau** [comment]
thanks for pinging this issue. The action item I had here was to, if possible, present an idea of how collaborators without access to the PUF could create a pull request that would run tests against the PUF but also not permit code that would subvert the system that keeps the PUF private.

If this is still important, which it sounds like it is, I have come across some relevant info that we could make use of in order to provide such a solution:

Travis CI permits one to specify private environment variables, which can be checked for during test execution (and then execute whatever logic is desired depending on the value of the variable). That part was already known. What is new is that I have discovered that these environment variables are only given a value when the pull request is initiated within a branch of this repo, not on a fork of this repo. The intention in this feature is that anyone who is trusted to push a branch to a given repo, is trusted with the environment variable values for the Travis CI setup on that repo. Otherwise, their Travis CI session proceeds with those variables as null. So, as an example, my typical pull request looks like:

```
open-source-economics/Tax-Calculator:master  <-----------    talumbau/Tax-Calculator:feature_branch
```

where the branch on the right contains my new code in my forked repo and I desire to merge it into master on the left. This type of pull request (which anyone may do) would have empty values for Travis CI environment variables. I suspect this is the workflow of many team members. The type of pull request that would actually run PUF-required tests would look like:

```
open-source-economics/Tax-Calculator:master  <-----------    open-source-economics/Tax-Calculator:feature_branch
```

So, to open such a pull request, one would actually push their branch to open-source-economics/Tax-Calculator instead of the fork.

The system would then work as follows:

 - @MattHJensen and others would designate who has push access to this repo
 - I (and perhaps others) would write some code that directs Travis CI to download the PUF each time the tests are run, using credentials kept in a environment variable on Travis CI
 - we turn the `requires_puf` tests on Travis CI
 - new collaborators, as well as anyone using the "PRs from forked repos" method for pull requests would see failures because Travis CI would attempt to download the PUF, fail, and then run the "PUF required" tests anyway and get a failure
 - collaborators who are trusted with push access to the repo would do all future PRs with the "push branch to upstream, not my fork" style. Travis CI would succeed at downloading the PUF and the tests would run as normal
 - for new collaborators whose code may need to run against the PUF before merging to master, a "core" member could look at their PR, make sure it is reasonable and not malicious code, then pull their branch into their own local repo, then push to upstream and Travis CI would run as normal.

Does this make sense and would it be useful to add?

**Comment by feenberg** [comment]
On Mon, 15 Feb 2016, T.J. Alumbaugh wrote:

> 
> thanks for pinging this issue. The action item I had here was to, if possible, present
> an idea of how collaborators without access to the PUF could create a pull request
> that would run tests against the PUF but also not permit code that would subvert the
> system that keeps the PUF private.
> 
> If this is still important, which it sounds like it is, I have come across some
> relevant info that we could make use of in order to provide such a solution:
> 
> Travis CI permits one to specify private environment variables, which can be checked
> for during test execution (and then execute whatever logic is desired depending on the
> value of the variable). That part was already known. What is new is that I have
> discovered that these environment variables are only given a value when the pull
> request is initiated within a branch of this repo, not on a fork of this repo. The
> intention in this feature is that anyone who is trusted to push a branch to a given
> repo, is trusted with the environment variable values for the Travis CI setup on that
> repo. Otherwise, their Travis CI session proceeds with those variables as null. So, as
> an example, my typical pull request looks like:
> 
> open-source-economics/Tax-Calculator:master  <-----------    talumbau/Tax-Calculator:f
> eature_branch
> 
> where the branch on the right contains my new code in my forked repo and I desire to
> merge it into master on the left. This type of pull request (which anyone may do)
> would have empty values for Travis CI environment variables. I suspect this is the
> workflow of many team members. The type of pull request that would actually run
> PUF-required tests would look like:
> 
> open-source-economics/Tax-Calculator:master  <-----------    open-source-economics/Tax
> -Calculator:feature_branch
> 
> So, to open such a pull request, one would actually push their branch to
> open-source-economics/Tax-Calculator instead of the fork.
> 
> The system would then work as follows:
>
>  *  @MattHJensen and others would designate who has push access to this repo
>  *  I (and perhaps others) would write some code that directs Travis CI to download
>     the PUF each time the tests are run, using credentials kept in a environment
>     variable on Travis CI
>  *  we turn the requires_puf tests on Travis CI
>  *  new collaborators, as well as anyone using the "PRs from forked repos" method for
>     pull requests would see failures because Travis CI would attempt to download the
>     PUF, fail, and then run the "PUF required" tests anyway and get a failure
>  *  collaborators who are trusted with push access to the repo would do all future PRs
>     with the "push branch to upstream, not my fork" style. Travis CI would succeed at
>     downloading the PUF and the tests would run as normal
>  *  for new collaborators whose code may need to run against the PUF before merging to
>     master, a "core" member could look at their PR, make sure it is reasonable and not
>     malicious code, then pull their branch into their own local repo, then push to
>     upstream and Travis CI would run as normal.
> 
> Does this make sense and would it be useful to add?

I don't have any experience with the test code.

After the tests run, who sees the output? What is in the output? I don't 
understand how run-time results wouldn't be a confidentiality problem if 
real data was used. Is real data necessary?

dan

**Comment by martinholmer** [comment]
T.J. (@talumbau),
Thanks for your extensive proposal, but I'm not sure it will dispel all of Matt's legitimate concerns about safeguarding the privacy of the IRS PUF data.  I'll let @MattHJensen decide that.  [I see Dan @feenberg has expressed these concerns while I was writing this response.]  For myself, I'm happy with the current situation, in which core developer (who have access to ```puf.csv``` run the ```requires_puf_csv``` tests on their local computer.

**Comment by MattHJensen** [comment]
@talumbau, I believe this would dispel my privacy concerns because only core developers would be able to trigger tests on GitHub that use the puf. I also agree with @martinholmer that the current situation is working out pretty well right now. Let's keep this option in mind and wait to see how the current workflow scales as the contributor community expands. We very well may want to circle back to it later, but for now, closing.
